File: Week_02-哈希表-集合-树-图/94.二叉树的中序遍历.py
#
# @lc app=leetcode.cn id=94 lang=python
#
# [94] 二叉树的中序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        """
        #方法1:递归
        #前序：根、左、右，即打印（输出）、左、右
        #中序：左、根、右，即左、打印（输出）、右
        #后序：左、右、根，即左、右、打印（输出）
        ans = list()
        #定义递归函数
        def dfs(root):
            #不满足条件返回上一层
            if not root:
                return
            #进入下一层
            dfs(root.left)
            #执行当前层动作，打印/输出
            ans.append(root.val)
            #进入下一层
            dfs(root.right)
        #进入递归函数
        dfs(root)
        #返回递归结果
        return ans
        """
        
        #方法2:非递归方法，迭代实现，使用栈（列表list）
        ''' 
        #代码风格1
        ans = list()
        stack = list()
        while stack or root:
            # 不断往左子树方向走，每走一次就将当前节点保存到栈中
			# 这是模拟递归的调用
            if root:
                stack.append(root)
                root = root.left
            # 当前节点为空，说明左边走到头了，从栈中弹出节点并保存
			# 然后转向右边节点，继续上面整个过程
            else:
                tmp = stack.pop()
                ans.append(tmp.val)
                root = tmp.right
        return ans
        '''
        #代码风格2
        ans = list()
        stack = list()
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            ans.append(root.val)
            root = root.right
        return ans

# @lc code=end

File: Week_02-哈希表-集合-树-图/94.二叉树的中序遍历.cpp
/*
 * @lc app=leetcode.cn id=94 lang=cpp
 *
 * [94] 二叉树的中序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include<vector>
using namespace std;
class Solution {
public:
#if 0


#if 1
//标准的递归写法
    void inorder(TreeNode* root, vector<int>& res) {
        //退出当前层
        if(!root) {
            return;
        }
        //进入到下一层
        inorder(root->left, res);
        //处理当前层逻辑
        res.push_back(root->val);
        //进入到下一层
        inorder(root->right, res);
    }
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        //进入递归的入口
        inorder(root, res);
        return res;
    }
#endif
    /*
    栈迭代方法思路：
    中序遍历
    每到一个节点 A，因为根的访问在中间，将 A 入栈。
    然后遍历左子树，接着访问 A，最后遍历右子树。
    在访问完 A 后，A 就可以出栈了。因为 A 和其左子树都已经访问完成。
    伪代码：
    初始化根节点
    枚举更新根节点，来自栈中暂存各级根节点的右子树
        枚举更新根节点，来自上次循环中根节点的左子树
            栈中暂存右子树；
        根节点指向当前根的左子树
    根节点指向栈顶暂存根节点的右子树，弹栈顶；
    输出根节点的值；
    更新根节点为栈中弹出根节点的右子树
    */
    vector<int> inorderTraversal(TreeNode* root) {
        stack<TreeNode*> st;
        vector<int> ans;
        while(root != nullptr || !st.empty()){
            while(root!=nullptr){
                st.push(root);
                root = root->left;
            }
            root = st.top();
            st.pop();
            ans.push_back(root->val);
            root = root->right;
        }
        return ans;        
    }
};
// @lc code=end

